js/data/feature_tree.js

```diff
@@ -1,73 +1,137 @@
 'use strict';
 
-var rbush = require('rbush');
 var Point = require('point-geometry');
-var util = require('../util/util');
 var loadGeometry = require('./load_geometry');
-var CollisionBox = require('../symbol/collision_box');
 var EXTENT = require('./bucket').EXTENT;
 var featureFilter = require('feature-filter');
+var StructArrayType = require('../util/struct_array');
+var Grid = require('../util/grid');
+var StringNumberMapping = require('../util/string_number_mapping');
+var vt = require('vector-tile');
+var Protobuf = require('pbf');
+var GeoJSONFeature = require('../util/vectortile_to_geojson');
+
+var FeatureIndexArray = new StructArrayType({
+    members: [
+        // the index of the feature in the original vectortile
+        { type: 'Uint32', name: 'featureIndex' },
+        // the source layer the feature appears in
+        { type: 'Uint16', name: 'sourceLayerIndex' },
+        // the bucket the feature appears in
+        { type: 'Uint16', name: 'bucketIndex' }
+    ]});
+
+var FilteredFeatureIndexArray = new StructArrayType({
+    members: [
+        // the index of the feature in the original vectortile
+        { type: 'Uint32', name: 'featureIndex' },
+        // the source layer the feature appears in
+        { type: 'Uint16', name: 'sourceLayerIndex' },
+        // the bucket the feature appears in
+        { type: 'Uint16', name: 'bucketIndex' },
+        // the layer the feature appears in
+        { type: 'Uint16', name: 'layerIndex' }
+    ]});
 
 module.exports = FeatureTree;
 
 function FeatureTree(coord, overscaling, collisionTile) {
+    if (coord.grid) {
+        var serialized = coord;
+        var rawTileData = overscaling;
+        coord = serialized.coord;
+        overscaling = serialized.overscaling;
+        this.grid = new Grid(serialized.grid);
+        this.featureIndexArray = new FeatureIndexArray(serialized.featureIndexArray);
+        this.rawTileData = rawTileData;
+        this.numberToLayerIDs = serialized.numberToLayerIDs;
+    } else {
+        this.grid = new Grid(16, EXTENT, 0);
+        this.featureIndexArray = new FeatureIndexArray();
+    }
+    this.coord = coord;
+    this.overscaling = overscaling;
     this.x = coord.x;
     this.y = coord.y;
     this.z = coord.z - Math.log(overscaling) / Math.LN2;
-    this.rtree = rbush(9);
-    this.toBeInserted = [];
     this.setCollisionTile(collisionTile);
 }
```

Comment:
The implementation I had in mind in https://github.com/mapbox/mapbox-gl-js/pull/2224#issuecomment-197043523 was more like:

```js
function FeatureTree(options) {
    this.overscaling = options.overscaling;
    this.rawTileData = options.rawTileData;
    this.numberToLayerIDs = options.numberToLayerIDs;
    this.coord = options.coord;
    
    this.featureIndexArray = new FeatureIndexArray(options.featureIndexArray);
    this.grid = new Grid(options.grid || {n: 15, extent: EXTENT, padding: 0});
    this.rtree = rbush(9);

    this.toBeInserted = [];

    this.x = this.coord.x
    this.y = this.coord.y
    this.z = this.coord.z - Math.log(overscaling) / Math.LN2;

    this.setCollisionTile(options.collisionTile);
}
```

Instead of thinking about the constructor of having "unserialized" mode and "normal" mode, think about it as accepting an `options` object flexible enough to hit both use cases.